url594c007908_hugobraga_College_tgz-pJ8-main_GraphJ8/src/algorithms/ApproximationAlgorithm.java

```diff
@@ -0,0 +1,292 @@
+        	qtTerminalsLeft = qtTerminalsLeft - (int)Math.floor(sqrtT);
+        	if (qtTerminalsLeft <= sqrtT)
+        		break;
+    	}
+
+    	uninformedAdjMatrix.refillMatrix();
+    	uninformedAdjMatrix.minCostPaths();
+    	
+    	qtTerminalsLeft = 0;
+    	for (int z = 0; z < terminals.length; z++){
+    		if (!informedSet.contains(terminals[z])) {
+    				terminalsLeft[z] = terminals[z];
+    				qtTerminalsLeft++;
+    		} else
+    			terminalsLeft[z] = -1;
+    	}    	
+    	
+    	ArrayList<ElkinItem> v2Set = new ArrayList<ElkinItem>(qtTerminalsLeft);
+    	
+    	uninformedTerminals = new ArrayList<Integer>(qtTerminalsLeft);
+    	uninformedTerminalsDelay = new ArrayList<Double>(qtTerminalsLeft);
+    	
+    	int j = 0;
+    	for (int i = 0; i < terminalsLeft.length; i++) {
+    		if (terminalsLeft[i] != -1) {//the value different from -1 indicates that the terminal is in UT
+    			uninformedTerminals.add(terminalsLeft[i]);
+    			uninformedTerminalsDelay.add(delaysLeft[i]);
+    			v2Set.add(new ElkinItem(null, new Node(terminalsLeft[i])));
+    		}
+    	}
+    	
+    	//finding out dominating set D
+    	applyDistDomProcedure(informedSet, v2Set);
+    	addingTreesForDomSet(uninformedTerminals);
+    	
+    	//connecting source to Roots
+    	P = globalAdjMatrix.getP2PPath();
+    	MCPmatrix = globalAdjMatrix.getP2PCost();
+    	for (int c = 0; c < roots.size(); c++) {
+    		int prev = roots.get(c);
+    		while (P[source][prev] != source) {
+    			quasiArborescence.checkInsertEdge(P[source][prev], prev, MCPmatrix[P[source][prev]][prev]);
+    			prev = P[source][prev];
+    		}
+    		quasiArborescence.checkInsertEdge(P[source][prev], prev, MCPmatrix[P[source][prev]][prev]);
+    	}
+    	
+    	//this procedure is necessary to guarantee that the final graph is a tree
+    	arborescence = buildSPT2Terminals(uninformedAdjMatrix, quasiArborescence, terminals);
+    	
+    	uninformedAdjMatrix.fillMatrixFromAdjList(arborescence);
+		uninformedAdjMatrix.minCostPaths();	
+    	return true;
+    }
+
+	@Override
+	public String getName() {
+		// TODO Auto-generated method stub
+		return new String("ElkinVariation");
+	}
+	
+	private void setFinalMatrix(AdjList arb) {
+		if (!finalMatrix) {
+			uninformedAdjMatrix.fillMatrixFromAdjList(arb);
+			uninformedAdjMatrix.minCostPaths();	
+			finalMatrix = true;
+		}
+	}	
+	
+	public double getTotalTerminalsCVR(double minCVR[], double maxCVR[]) {
+		setFinalMatrix(arborescence);
+		
+		double terFinalCosts[] = getCosts(uninformedAdjMatrix, terminals);
+		double terSpannerCosts[] = getSpannerQuant();
+		//double min, max;
+		
+		minCVR[0] = 0;
+		maxCVR[0] = 0;
+		double ratios = 0;
+		for (int i = 0; i < terFinalCosts.length; i++) {
+			double val = terFinalCosts[i]/terSpannerCosts[i];
+			if (val > maxCVR[0])
+				maxCVR[0] = val;
+			if ((val < minCVR[0]) || (minCVR[0] == 0))
+				minCVR[0] = val;
+			ratios += terFinalCosts[i]/terSpannerCosts[i];
+		}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 172:179, 204:217, 245:253.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.